File: src/test/java/org/goafabric/personservice/persistence/extensions/AuditTrailListenerIT.java
package org.goafabric.personservice.persistence.extensions;

import jakarta.persistence.EntityManager;
import jakarta.persistence.PersistenceContext;
import org.goafabric.personservice.controller.PersonController;
import org.goafabric.personservice.controller.dto.Address;
import org.goafabric.personservice.controller.dto.Person;
import org.goafabric.personservice.persistence.PersonRepository;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;

import java.util.List;
import java.util.Objects;

import static org.assertj.core.api.Assertions.assertThat;

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
class AuditTrailListenerIT {
    @Autowired
    private PersonController personController;

    @Autowired
    private PersonRepository personRepository;

    @PersistenceContext
    private EntityManager entityManager;

    @Test
    public void creatUpdateDeletePerson() {
        var person = save();

        var createPerson = selectFrom("CREATE", person.id());
        assertThat(createPerson.getOldValue()).isNull();
        assertThat(createPerson.getNewValue()).isNotNull();
        assertThat(Objects.requireNonNull(createPerson.getNewValue()))
                .isNotNull().contains("Marge", "Simpson");

        var updatePerson = selectFrom("UPDATE", person.id());
        assertThat(updatePerson.getOldValue()).isNotNull();
        assertThat(updatePerson.getOldValue()).isNotNull();
        assertThat(Objects.requireNonNull(updatePerson.getOldValue()))
                .isNotNull().contains("Marge", "Simpson");
        assertThat(Objects.requireNonNull(updatePerson.getNewValue()))
                .isNotNull().contains("updatedFirstName", "updatedLastName");

        var deletePerson = selectFrom("DELETE", person.id());
        assertThat(deletePerson.getOldValue()).isNotNull();
        assertThat(deletePerson.getNewValue()).isNull();
        assertThat(Objects.requireNonNull(deletePerson.getOldValue()))
                .isNotNull().contains("updatedFirstName", "updatedLastName");
    }

    @Test
    public void creatUpdateDeleteAddress() {
        var address = save().address().getFirst();

        var createAddress = selectFrom("CREATE", address.id());
        assertThat(createAddress.getOldValue()).isNull();
        assertThat(createAddress.getNewValue()).isNotNull();
        assertThat(Objects.requireNonNull(createAddress.getNewValue()))
                .isNotNull().contains("Terrace");


        /*
        var updateAddress = selectFrom("UPDATE", address.id());
        assertThat(updateAddress.getOldvalue()).isNotNull();
        assertThat(updateAddress.getNewvalue()).isNotNull();

         */

        var deleteAddress = selectFrom("DELETE", address.id());
        assertThat(deleteAddress.getOldValue()).isNotNull();
        assertThat(deleteAddress.getNewValue()).isNull();
        assertThat(Objects.requireNonNull(deleteAddress.getOldValue()))
                .isNotNull().contains("Terrace");

    }

    private AuditTrailListener.AuditTrail selectFrom(String operation, String id) {
        var query = entityManager.createQuery(
                "SELECT a FROM AuditTrailListener$AuditTrail a WHERE a.objectId = :objectId AND a.operation = :operation", AuditTrailListener.AuditTrail.class);
        query.setParameter("objectId", id);
        query.setParameter("operation", AuditTrailListener.DbOperation.valueOf(operation));
        return query.getSingleResult();
    }

    Person save() {
        final Person person = personController.save(
                new Person(null,
                        null,
                        "Marge",
                        "Simpson",
                        List.of(
                                createAddress("Evergreen Terrace"),
                                createAddress("Everblue Terrace"))
                ));

        //update
        personController.save(new Person(person.id(), person.version(),
                "updatedFirstName", "updatedLastName", person.address()));

        personRepository.deleteById(person.id());
        return person;
    }

    private Address createAddress(String street) {
        return new Address(null, null,
                street, "Springfield");
    }

}

File: src/test/kotlin/org/goafabric/personservice/persistence/extensions/AuditTrailListenerIT.kt
package org.goafabric.personservice.persistence.extensions

import jakarta.persistence.EntityManager
import org.assertj.core.api.Assertions.assertThat
import org.goafabric.personservice.controller.PersonController
import org.goafabric.personservice.controller.dto.Address
import org.goafabric.personservice.controller.dto.Person
import org.goafabric.personservice.persistence.PersonRepository
import org.goafabric.personservice.persistence.extensions.AuditTrailListener.AuditTrail
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import java.util.*
import java.util.List

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
internal class AuditTrailListenerIT(
    @Autowired val personController: PersonController,
    @Autowired val personRepository: PersonRepository,
    @Autowired val entityManager: EntityManager) {

    @Test
    fun creatUpdateDeletePerson() {
        val person = save()

        val createPerson = selectFrom("CREATE", person.id)
        assertThat(createPerson.oldValue).isNull()
        assertThat(createPerson.newValue).isNotNull()
        assertThat(Objects.requireNonNull(createPerson.newValue))
            .isNotNull().contains("Homer", "Simpson")

        val updatePerson = selectFrom("UPDATE", person.id)
        assertThat(updatePerson.oldValue).isNotNull()
        assertThat(updatePerson.oldValue).isNotNull()
        assertThat(Objects.requireNonNull(updatePerson.oldValue))
            .isNotNull().contains("Homer", "Simpson")
        assertThat(Objects.requireNonNull(updatePerson.newValue))
            .isNotNull().contains("updatedFirstName", "updatedLastName")

        val deletePerson = selectFrom("DELETE", person.id)
        assertThat(deletePerson.oldValue).isNotNull()
        assertThat(deletePerson.newValue).isNull()
        assertThat(Objects.requireNonNull(deletePerson.oldValue))
            .isNotNull().contains("updatedFirstName", "updatedLastName")
    }

    @Test
    fun creatUpdateDeleteAddress() {
        val address = save().address.first()

        val createAddress = selectFrom("CREATE", address?.id)
        assertThat(createAddress.oldValue).isNull()
        assertThat(createAddress.newValue).isNotNull()
        assertThat(Objects.requireNonNull(createAddress.newValue))
            .isNotNull().contains("Terrace")


        /*
        var updateAddress = selectFrom("UPDATE", address.id());
        assertThat(updateAddress.oldValue).isNotNull();
        assertThat(updateAddress.newValue).isNotNull();

         */
        val deleteAddress = selectFrom("DELETE", address?.id)
        assertThat(deleteAddress.oldValue).isNotNull()
        assertThat(deleteAddress.newValue).isNull()
        assertThat(Objects.requireNonNull(deleteAddress.oldValue))
            .isNotNull().contains("Terrace")
    }

    private fun selectFrom(operation: String, id: String?): AuditTrail {
        val query = entityManager!!.createQuery<AuditTrail>(
            "SELECT a FROM AuditTrailListener\$AuditTrail a WHERE a.objectId = :objectId AND a.operation = :operation",
            AuditTrail::class.java
        )
        query.setParameter("objectId", id)
        query.setParameter("operation", AuditTrailListener.DbOperation.valueOf(operation))
        return query.getSingleResult()
    }

    fun save(): Person {
        val person = personController!!.save(
            Person(
                null,
                null,
                "Homer",
                "Simpson",
                List.of<Address?>(
                    createAddress("Evergreen Terrace"),
                    createAddress("Everblue Terrace")
                )
            )
        )

        //update
        personController.save(
            Person(
                person.id, person.version,
                "updatedFirstName", "updatedLastName", person.address
            )
        )

        personRepository!!.deleteById(person.id!!)
        return person
    }

    private fun createAddress(street: String): Address {
        return Address(
            null, null,
            street, "Springfield"
        )
    }
}